### PR TITLE
rtdl: implement RTLD_NEXT

### LINKED ISSUE #577: rtdl: RTLD_NEXT is unimplemented
While trying to use `libasan` on mlibc LFS, we ran into [this](https://github.com/managarm/mlibc/blob/a53280ab5fa0a851b62abcbf77bf0c41707199a1/options/rtdl/generic/main.cpp#L392) ensure, preventing asan from working. Fixing it would allow us to use `libasan`.

### PR DESCRIPTION
**Pull request by 64**
Fixes #577. Split off from #448. Blocked on the new RTDL testing framework in #586 so I can add a meaningful test for this.

### CODE REVIEW

File: options/posix/generic/dlfcn-stubs.cpp [hunk 1]
```diff
@@ -12,37 +12,27 @@ struct __dlapi_symbol {
...
-		mlibc::infoLogger() << "mlibc: unhandled dlopen flags " << unhandled << frg::endlog;
-
-	return __dlapi_open(file, !(flags & RTLD_GLOBAL));
+	auto ra = __builtin_extract_return_addr(__builtin_return_address(0));
+	return __dlapi_open(file, flags, ra);
 }
 
 void *dlsym(void *__restrict handle, const char *__restrict string) {
-	return __dlapi_resolve(handle, string);
+	auto ra = __builtin_extract_return_addr(__builtin_return_address(0));
```
Comment (avdgrinten) [inline options/posix/generic/dlfcn-stubs.cpp:34]: The functions containing this should be marked `[[gnu::noinline]]`.

File: options/rtdl/generic/linker.cpp [hunk 2]
```diff
@@ -262,6 +262,25 @@ SharedObject *ObjectRepository::requestObjectAtPath(frg::string_view path, uint6
...
+	uintptr_t target = reinterpret_cast<uintptr_t>(addr);
+
+	for (auto [name, object] : _nameMap) {
+		// Search all PT_LOAD segments for the specified address.
+		for(size_t j = 0; j < object->phdrCount; j++) {
+			auto phdr = (Elf64_Phdr *)((uintptr_t)object->phdrPointer + j * object->phdrEntrySize);
+			if (phdr->p_type == PT_LOAD) {
+				uintptr_t start = object->baseAddress + phdr->p_vaddr;
+				uintptr_t end = start + phdr->p_memsz;
+				if (start <= target && target <= end)
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:275]: `target < end`, right? For example, if start = 0x1000, size = 0x1000, then 0x2000 is already on the next page.

File: options/rtdl/generic/main.cpp [hunk 3]
```diff
@@ -385,16 +402,22 @@ void *__dlapi_open(const char *file, int local) {
...
 	if (handle == RTLD_DEFAULT) {
 		target = Scope::resolveWholeScope(globalScope.get(), string, 0);
+	} else if (handle == RTLD_NEXT) {
+		SharedObject *origin = initialRepository->findCaller(returnAddress);
+		if (!origin) {
+			mlibc::panicLogger() << "rtdl: unable to determine calling object of dlsym "
+				<< "(ra = " << returnAddress << ")" << frg::endlog;
+		}
+
+		target = Scope::resolveNext(globalScope.get(), string, origin);
```
Comment (avdgrinten) [inline options/rtdl/generic/main.cpp:420]: The local scope is not handled here (yet)?

Comment (64) [reply options/rtdl/generic/main.cpp:420]: Yeah, I guess we just use the object's scope when that is implemented

File: tests/rtdl/rtld_next/meson.build [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Prevent tail calls, as it breaks libc's 'calling object' detection.
+no_tail_calls = '-fno-optimize-sibling-calls'
```
Comment (avdgrinten) [inline tests/rtdl/rtld_next/meson.build:2]: Is this needed despite `[[gnu::noinline]]`?

**Review by avdgrinten — COMMENTED** [review]
LGTM except for the comment below.

File: tests/rtdl/rtld_next/meson.build (hunk 4 above)
Comment (64) [reply tests/rtdl/rtld_next/meson.build:2]: Yes:
```c
void *foo() {
    return dlsym(...);
}
```
Even though dlsym isn't inlined, the compiler will generate a tail call here. But that'd cause the return address visible to dlsym to be foo's caller instead, which is not what we want if foo is called from a different object.

**Pull request merged**